### ISSUE TITLE
Adding saving of aggregated base load

### ISSUE THREAD

**Issue by grdw** [quintel/etmoses#246]
If I'm correct there's a flexible and inflexible profile.

Thoughts:
```text
Currently both profiles are summed. I don't know how, but it seems that way.

This strategy removes the flexible part from the buildings curve.

There are two ways to do this: split the profiles somehow mold them into the network technologies thingy (???).

Or just remove the flex part with a negative `consumption` which seems a gazillion times easier though it'll be ugly. 
````

**Comment by antw** [comment]
> Currently both profiles are summed. I don't know how, but it seems that way.


[They're summed here](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/installed_technology.rb#L92-L106). This was a necessary stopgap solution to the fact that previously only the first curve component was being used.

> There are two ways to do this: split the profiles somehow mold them into the network technologies thingy (???).

Since the flexible and inflexible profiles are already separate, the proper solution in the network calculation is to represent them as two separate technologies. The inflexible part should be a [Technologies::Generic](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/network/technologies/generic.rb), and the flexible part is a [Technologies::OptionalConsumer](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/network/technologies/optional_consumer.rb). The network technology objects [are constructed here](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/calculation/technology_load.rb#L14-L16).

If the user does not select the "Saving of aggregated base-load" option in the front-end, the flexible technology will automatically do the right thing by becoming inflexible.

**Comment by antw** [comment]
To add to this: we probably need to detect how many curve components the technology has. If it has only one, then it is an "individual base load" and should be postponable.

##### 1 component:
* Technology is a DeferrableConsumer

##### 2 components:
* Inflexible part is a Generic
* Flexible part is an OptionalConsumer

**Comment by ChaelKruip** [comment]
Just to clarify: both aggregated base load and individual base load profiles have 2 curves, a flexible and an inflexible one. For the aggregated case, the flexible part can be saved/reduced. For the 'individual' base loads, the flexible part can be postponed.

**Comment by antw** [comment]
> For the aggregated case, the flexible part can be saved/reduced. For the 'individual' base loads, the flexible part can be postponed.

Oh, that's unfortunate, but it makes more sense. :laughing: 

Would be nice to see become separate technologies then: a `base_load_large` which uses the EDSN profiles and saving, and `base_load_small` which uses individual profiles and postponing. Not sure how easy that is to do?

**Comment by antw** [comment]
I added [a wiki page describing base loads](https://github.com/quintel/etmoses/wiki/Base-loads). Might be nice to expand on it later, such as:

1. What's the difference between an EDSN profile and an individual profile?
2. Why is an individual profile unsuitable for use with large nodes, and vice versa?
3. Why are small endpoints postponed, but large ones are "saved"?
4. How do I upload my own flexible and inflexible profiles and associate them with one another?

... but that's not urgent.

**Comment by grdw** [comment]
>Since the flexible and inflexible profiles are already separate...

They are not treated as such right now. They're summed in the piece of code above. If I understand correctly in dead-simple code:

````ruby
inflex_profile = [1,2,3]
flex_profile =   [1,2,3]

# the actual profile which is visible in the front-end
actual_profile = [2,4,6]
````

> ...The proper solution in the network calculation is to represent them as two separate technologies. The inflexible part should be a Technologies::Generic, and the flexible part is a Technologies::OptionalConsumer. The network technology objects are constructed here.

> If the user does not select the "Saving of aggregated base-load" option in the front-end, the flexible technology will automatically do the right thing by becoming inflexible.

What I don't get about this: is this already as such or does this need to be changed? And more concretly what needs to be changed? There's already something happening as I describe down below but I can't really find where in the code that is happening. What I understood what I should do is: cutting up `flex` and `inflex` - using separate technologies - assign those profiles and switching them with the `saving_base_load` option. However something is already happening and I don't know how/where/what? :disappointed_relieved: 

**How I see this feature being implemented right now**
If I tick the `saving of aggregated base load` button. The profile is "flatlined" until the capacity of the node (which is what I'm seeing) This is the case if you only have households in a testing grounds which is the case for the screenshot.
![screen shot 2015-07-29 at 9 41 22 am](https://cloud.githubusercontent.com/assets/2676542/8952292/0024d6da-35d6-11e5-9017-3a5d03221c27.png)

**What I've done so far**
- Changed the load profile uploading that it's possible to upload `use` and `availibility` profiles as separate `load_profile_components`
- Trying to wrap my head around the current changes made in the `app/models/network`- folder.
- Writing specs for what should happen if the user selects a certain strategy from the list.

**TL;DR**
What I'm actually aking here: is the `saving of aggregeted base load` already working? Which parts are, which parts aren't, what actually needs changing. I'm not really asking about `postponing` here because I'm not really focussing on that in this issue (will use another issue for that).

**Comment by grdw** [comment]
I wrote specs so I found out that this:

````ruby
inflex_profile = [1,2,3]
flex_profile =   [1,2,3]

# the actual profile which is visible in the front-end
actual_profile = [2,4,6]
````

is actually this:

````ruby
inflex_profile = [1,2,3]
flex_profile =   [1,2,3]

# the actual profile which is visible in the front-end
actual_profile = [1,2,3,1,2,3]
````

**Comment by antw** [comment]
> I wrote specs so I found out that this: [...] is actually this: [...]

This is why stubs are dangerous. :smile:

The test [stubs InstalledTechnology to return an `Array`](https://github.com/quintel/etmoses/commit/4bcee27563d618a75abc18a4e15a3db6dd377c24#diff-697edaa1d8f85247b43b4350cd51e35bR31). The real implementation returns a `Network::Curve`. There are several important differences; in this case:

```ruby
include Network

arr1 = [1.0, 2.0]
arr2 = [3.0, 4.0]

# Summing arrays:
arr1 + arr2 # => [1.0, 2.0, 3.0, 4.0]

# Summing curves:
(Curve.new(arr1) + Curve.new(arr2)).to_a # => [4.0, 6.0]
```

> What I don't get about this: is this already as such or does this need to be changed?

For base loads, what we have right now is a temporary solution:

* We combine the flexible and inflexible curves ([sum them together and scale them](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/installed_technology.rb#L97-L105) so that the maximum value in the combined curve is 1.0).
* From this, we create a single network technology. If the user selects "saving of aggregated base load", the entire technology becomes flexible, and the load can be "saved" (the technology is an instance of Technologies::OptionalConsumer).

In your example above, the technology "flatlines" at the network capacity because the entire base load is being treated as flexible.

The *correct* behaviour is not to create a single network technology for the base load, but two:

* The inflexible part should be a Technologies::Generic.
* The flexible part should be a Technologies::OptionalConsumer ("saving") or Technologies::DeferredConsumer ("postponing").

I think the place to do this is [Calculation::TechnologyLoad#run](https://github.com/quintel/etmoses/blob/master/app/models/calculation/technology_load.rb#L11-L21). In pseudo-code:

```ruby
node.set(:techs, suitable_technologies(node).flat_map do |tech|
  if tech is a base load
    # This branch is hypothetical: InstalledTechnology#each_component doesn't exist,
    # nor does Technologies.from_component.
    tech.each_component.map do |component|
      Network::Technologies.from_component(
        tech, component, profile_for(component), @context.options
      )
  else
    Network::Technologies.from_installed(
      tech, profile_for(tech), @context.options
    )
  end
end)
```

> assign those profiles and switching them with the `saving_base_load` option

You don't need to swap technologies depending on the user options, the network model will do this for you automatically. We build the network as if all the features are turned on and [Disableable](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/network/technologies/disableable.rb) inspects the user options and configures the technologies appropriately.

For example, if you try to build an OptionalConsumer, but `saving_base_load` is turned off, you'll actually get a [Generic](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/network/technologies/generic.rb) back instead. [See these lines](https://github.com/quintel/etmoses/blob/a8dd8146afb361328226d11a5221010b357a9eb5/app/models/network/technologies/optional_consumer.rb#L10-L16). If this is still confusing, don't worry too much about it for now: just know that if you try to create an optional or deferred consumer, and the user has disabled those technologies, the network will do the right thing.

**Comment by grdw** [comment]
> This is why stubs are dangerous. :smile:

Let me just change that to: that's why *wrong* stubs are ~~dangerous~~ stupid.

Thanks for the explanation though! This helps a lot.

**Comment by ChaelKruip** [comment]
Github is much more interesting than the novel I'm trying to read here on the beach ;)

Just one clarification/correction to the excellent explanation of @antw: the base load profiles are not scaled to have a maximum of 1.0. Instead, they are scaled such that their integral is equal to the total electricity demand of the household or building. I think this is done correctly at the moment though. A simple check is to sum all values of the resulting curve. This should be equal to the demand in kWh times four (the factor of four is due to the fact that there are four quarters in an hour).

**Comment by antw** [comment]
> Instead, they are scaled such that their integral is equal to the total electricity demand of the household or building.

Yes! Apologies. Base loads have a `demand` rather than a `capacity` or `load`. The former is scaled so that the sum of all points is 1.0, the latter so that the maximum value is 1.0.

* Input: `[0.1, 0.4]`
* Capacity-scaled: `[0.25, 1.0]`  (maximum single value is 1.0)
* Demand-scaled: `[0.2, 0.8]`  (sum of all values is 1.0)

**Comment by grdw** [comment]
The `:demand_scaled` scaling scales to '1.0'. However this is not really practical for `flex` and `inflex` due to the ratio between the two profiles of `[0.9, 0.1]` (is this the correct ratio?)(will it by dynamic in the future?)

One simple solution would be to have a hash:

````ruby
SCALING_VALUES = {
   flex: 0.9,
   inflex: 0.1,
   default: 1
}
````

And multiply it [here](https://github.com/quintel/etmoses/blob/issues/load-management/app/models/installed_technology.rb#L91).

**Comment by grdw** [comment]
Nevermind. Joris and I decided to calculate the ratio between the two original profiles and use that ratio to create the correct curve.
